core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -175,6 +178,22 @@ abstract class AbstractFetcherThread(name: String,
     }
   }
 
+  private def doTruncate(topicPartition: TopicPartition, truncationState: OffsetTruncationState): Unit = {
+    try {
+      truncate(topicPartition, truncationState)
+    }
+    catch {
+      case e: KafkaStorageException =>
```

Comment:
Storage failures result in an event which eventually bubbles up to `ReplicaManager.handleLogDirFailure` by way of `LogDirFailureChannel`. This method will remove the partitions from the replica fetchers, which will consequently remove them from the `failedPartitions` set. I am wondering if it would be better not to do this so that `FailedPartitionsCount` also takes into account failed log directories.

cc @junrao Does this make sense?

Reply:
Thinking about this a little more, we already have a separate metric to track the offline partitions in `ReplicaManager`, so I think we do not need to change the behavior of `handleLogDirFailure`. If a partition fails due to a storage error, it will be reflected initially in `FailedPartitionsCount`, but later in the offline metric. I think that seems reasonable.